subprojects/core/src/main/java/org/gradle/api/internal/plugins/SoftwareTypeRegistrationPluginTarget.java

```diff
@@ -81,23 +87,56 @@ private void registerSoftwareTypes(TypeMetadata typeMetadata) {
         Optional<RegistersSoftwareTypes> registersSoftwareType = typeMetadata.getTypeAnnotationMetadata().getAnnotation(RegistersSoftwareTypes.class);
         registersSoftwareType.ifPresent(registration -> {
             for (Class<? extends Plugin<Project>> softwareTypeImplClass : registration.value()) {
-                validateSoftwareTypePluginExposesSoftwareTypes(softwareTypeImplClass);
+                validateSoftwareTypePluginExposesExactlyOneSoftwareType(softwareTypeImplClass, typeMetadata.getType());
                 softwareTypeRegistry.register(softwareTypeImplClass);
             }
         });
     }
 
-    void validateSoftwareTypePluginExposesSoftwareTypes(Class<? extends Plugin<Project>> softwareTypePluginImplClass) {
+    void validateSoftwareTypePluginExposesExactlyOneSoftwareType(Class<? extends Plugin<Project>> softwareTypePluginImplClass, Class<?> registeringPlugin) {
+        DefaultTypeValidationContext typeValidationContext = DefaultTypeValidationContext.withRootType(softwareTypePluginImplClass, false);
         TypeToken<?> softwareTypePluginImplType = TypeToken.of(softwareTypePluginImplClass);
         TypeMetadata softwareTypePluginImplMetadata = inspectionScheme.getMetadataStore().getTypeMetadata(softwareTypePluginImplType.getRawType());
+        softwareTypePluginImplMetadata.visitValidationFailures(null, typeValidationContext);
 
-        for (PropertyMetadata propertyMetadata : softwareTypePluginImplMetadata.getPropertiesMetadata()) {
-            Optional<SoftwareType> registersSoftwareType = propertyMetadata.getAnnotation(SoftwareType.class);
-            if (registersSoftwareType.isPresent()) {
-                return;
-            }
+        long exposedSoftwareTypes = softwareTypePluginImplMetadata.getPropertiesMetadata().stream()
```

Comment:
💭  
```
List<String> registeredSoftwareTypes = softwareTypePluginImplMetadata.getPropertiesMetadata().stream()
            .map(propertyMetadata -> propertyMetadata.getAnnotation(SoftwareType.class))
            .filter(Optional::isPresent)
            .map(annotation -> annotation.get().name())
            .sorted()
            .collect(Collectors.toList());
```

And then have the list of names to put in the error message, which can also be counted.

Reply:
Done.